### DISCUSSION TITLE
Non-nullable types with z.ZodType<T> ?

### DISCUSSION THREAD

**Discussion by torjue** [colinhacks/zod#3357] [Q&A]
I'm trying to get ZodType to work with booleans. If I use `z.ZodType<T>` I need to have the boolean property as a nullable boolean to avoid the following error:

 ```typescript
import z from "zod";

export type NonWorkingType = {
	someBoolean: boolean;
};

export const nonWorkingSchema: z.ZodType<NonWorkingType> = z.object({
	someBoolean: z.boolean().default(false),
});
// Type 'ZodObject<{ someBoolean: ZodDefault<ZodBoolean>; }, "strip", ZodTypeAny, { someBoolean: boolean; }, { someBoolean?: boolean | undefined; }>' is not assignable to type 'ZodType<NonWorkingType, ZodTypeDef, NonWorkingType>'.
//   The types of '_input.someBoolean' are incompatible between these types.
//     Type 'boolean | undefined' is not assignable to type 'boolean'.
//       Type 'undefined' is not assignable to type 'boolean'
```

(Adding [`.required`](https://zod.dev/?id=required) doesn't help, either.)

As you can see, it works if the boolean is nullable:

```typescript
import z from "zod";

export type WorkingType = {
	someBoolean?: boolean;
};

export const workingSchema: z.ZodType<WorkingType> = z.object({
	someBoolean: z.boolean().default(false),
});
```

On the other hand, if I don't use `z.ZodType<T>`, and use `z.infer<>` instead, I get the correct type with a not nullable boolean:

```typescript
import z from "zod";

export const inferedSchema = z.object({
	someBoolean: z.boolean().default(false),
});

type InferedType = z.infer<typeof inferedSchema>;
// type InferedType = {
//     someBoolean: boolean;
// }
```

(Adding [`.partial`](https://zod.dev/?id=partial) creates a nullable type, as expected.)

Anything I'm missing? Is it expected behaviour, or a bug? 

Using `z.ZodType<T>` is my preferred way as it behaves better with Visual Studio Code, as seen in [this StackOverflow post](https://stackoverflow.com/questions/74068609/zod-show-inferred-nested-types-in-ide).

**Comment by torjue** [comment]
After some more research I see that it only happens with `.default(boolean)`:

Works:

```typescript
import z from "zod";

export type WorkingType = {
	someBoolean: boolean;
};

export const workingSchema: z.ZodType<WorkingType> = z.object({
	someBoolean: z.boolean(),
});
```

Doesn't work:
```typescript
import z from "zod";

export type NonWorkingType = {
   someBoolean: boolean;
};

export const nonWorkingSchema: z.ZodType<NonWorkingType> = z.object({
   someBoolean: z.boolean().default(false),
});
```

**Reply by MaLiN2223** [reply]
Did you ever find the solution?

**Comment by ilyazub** [comment]
This is what worked for me:

```ts
preprocess((val) => String(val).toLowerCase() === 'true', z.boolean()).default(false)
```

Thanks to https://github.com/colinhacks/zod/issues/2985#issuecomment-2085239542. (@IlmariKu thanks!)
